### DISCUSSION TITLE
Usage of ConvexMeshShape with collider

### DISCUSSION THREAD

**Discussion by Bubbasm** [DanielChappuis/reactphysics3d#264] [Q&A] [answered]
I am getting difficulties with creating a convex mesh and adding a collider to it. 

I have taken the example straight from the manual (https://www.reactphysics3d.com/usermanual.html#x1-5100011.1.4), but I am not getting the expected result.

The segment of code where I try to add the convexMeshShape of a cube is the following:
 ``` 
auto transformObject = rp3d::Transform{
      SvkeUtils::convertRP3D(objManager.transformComponent[transformId].translation),
      SvkeUtils::convertRP3D(objManager.transformComponent[transformId].quaternion)};
  physicsBodies[id] = world->createRigidBody(transformObject);
  physicsBodies[id]->setLinearVelocity(SvkeUtils::convertRP3D(velocity));
  if (!moovable) {
    physicsBodies[id]->setType(rp3d::BodyType::STATIC);
  } else {
    physicsBodies[id]->setType(rp3d::BodyType::DYNAMIC);
  }
  physicsBodies[id]->addCollider(convexMeshShape, rp3d::Transform::identity());
```

The expected  outcome should be a cube with dimensions 6x6x6.

The result I am getting is that the cube is not interacting appropriately with other objects in the scene, such as spheres and boxes created with `physicsCommon.createSphereShape()` and `physicsCommon.createBoxShape()` accordingly. Note that objects created from the latter shape creators do work flawlessly.

**Comment by DanielChappuis** [comment]
Hello,

I am not sure this is the issue but do not forget to compute or set a correct inertia tensor for your body as described in the documentation [here](https://www.reactphysics3d.com/usermanual.html#x1-4300010.8.3). For instance, you can call the `RigidBody:: updateMassPropertiesFromColliders()` method to automatically compute an inertia tensor.

If it is not better, you can try to take a look at [this code](https://github.com/DanielChappuis/reactphysics3d/blob/master/testbed/common/ConvexMesh.cpp) in the testbed application. That's the code that create a ConvexMeshShape. Maybe it will help to figure out what is wrong.

Also note that you can use the [DebugRenderer](https://www.reactphysics3d.com/usermanual.html#x1-8100018) of the library to display the faces of your ConvexMeshShape in order to see if everything is correct.

**Reply by Bubbasm** [reply]
After a while trying to debug what is wrong, I finally used the DebugRenderer, and it seems as the collision shape of the object created by a ConvexMeshShape is not correct:

![Screenshot_20220804_132744](https://user-images.githubusercontent.com/26069774/182847175-cb6e785d-0808-495c-90d3-02cc424a2f08.png)

And I really have no idea what is going on, as I literally copied and pasted the [ConvexMeshShape](https://www.reactphysics3d.com/usermanual.html#x1-5100011.1.4) creation from the manual.

I have also added the `updateMassPropertiesFromColliders` and `updateLocalIntertiaTensorFromColliders` but it does not help.
I should also mention that I have used the reactphysics3d library compiling it with both single precision and double precision, but the end result is the same.

**Comment by DanielChappuis** [comment, marked as answer]
I took the time to try again with the code from the documentation. I replaced the code that creates a ConvexMeshShape in the testbed application [here](https://github.com/DanielChappuis/reactphysics3d/blob/master/testbed/common/ConvexMesh.cpp) by the code from the documentation that creates a simple cube and it's working fine. If I display the collision shapes in the tesbed application, I can see that the collision shape is now a cube.

Can you check in your code that your your array of **vertices**, **indices**, **PolygonFace**, your **PolygonVertexArray** and **PolyhedronMesh** are persisted in memory and not destroyed right after the ConvexMeshShape creation. As written in the [documentation ](https://reactphysics3d.com/usermanual.html#x1-5100011.1.4)(see quote below), those objects needs to persist in memory during **the whole simulation** not just during collision shape creation.

> From documentation: Note that the vertex coordinates and indices array are not copied and therefore you need to make sure that they exist until the collision shape exists. This is also true for the all the PolygonFace, the PolygonVertexArray and the PolyhedronMesh objects.

**Reply by Bubbasm** [reply]
I just realized this today, while debugging with the code on github. And indeed, that was the mistake.

As per the quote, I initially misunderstood it. What I understood is that all those objects (vertices, indices, PolygonFace, PolygonVertexArray and PolyhedronMesh) needed to exist in memory only until the collision shape is **created** (e.g. calling `addCollider` method of a RigidBody). Now I realize that it is for the entire existance of the collision shape.

Thank you very much for your time :)
